[MRG] Changed examples so they produce the same values on OS X

georgipeev opened the pull request:
#### Reference Issues/PRs
Fixes #10213.

#### What does this implement/fix? Explain your changes.
Doctests were failing because of different math libraries used for random number generation. Decreasing the tolerance fixed the problem. Also switched the scoring method for cross validation and the corresponding results.

Commit:
changed examples so they produce the same values on MacOS

amueller commented:
lgtm if it passes ;)

twosigmajab committed:
DOC add missing requirements for building docs (#11292)

Fixes #11286.

TomDLT commented:
Almost there, fails only on one build:
```
    -     random_state=0, tol=0.00001, verbose=0)
    +     random_state=0, tol=1e-05, verbose=0)
```
```
    -     multi_class='ovr', penalty='l2', random_state=0, tol=0.00001,
    -     verbose=0)
    +     multi_class='ovr', penalty='l2', random_state=0, tol=1e-05, verbose=0)
```

tsdlovell committed:
DOC: replace TODO with link to the glossary (#11279)

Fixes #9294

amueller commented:
So that's a numpy version thing? Should we add a fixture? seems a bit overkill... this doesn't happen in other places?

Andrew-peng committed:
FIX #11215 : Changing return in docstring to yields for generator functions (#11276)

georgipeev commented:
Doc tests pass even if I change those 0.00001s to scientific notation. I'll do that.

Commit:
switched two constants to scientific notation

twosigmajab committed:
DOC replace OpenHub/ohloh badge with star button (#11288)

aozgaa committed:
MAINT skip dataset downloading doctest (#11284)

agramfort committed:
DOC: add references for CD in LASSO and duality gap criterion (#11302)

scottgigante committed:
Add sparse efficiency warning to randomized_svd for dok_matrix / lil_matrix (#11264)

thomasjpfan committed:
FIX Uses self.scoring for score function (#11192)

agramfort committed:
BLD fix sphx gallery errors (#11307)

yufengg committed:
proposal to use .joblib file extension (#11230)

Recommend to use of `filename.joblib` instead of `filename.pkl` for models persisted via the joblib library to reduce confusion when it comes to time load a model, as it will be more clear whether a file was saved using the `pickle` or `joblib` library.

lesteve committed:
DOC: use .joblib file extension rather than .pkl

Commit:
use slipsis for example values that depend on RNG library

amueller commented:
That seems like a reasonable level of precision for adding ellipsis, though :)

qinhanmin2014 commented:
Personally, I don't like making examples complex (e.g., using something like ``scoring='recall_macro'``, ``LinearSVC(tol=0.00001)`` ). I think it's unfriendly to users, especially new comers. I wonder whether it's better to solve the problem by only reducing decimal digits (e.g., ``16.35750999`` -> ``16.35``).

Commit:
addede more ellipsis

amueller commented:
The point of the cross-validation example was to show off the scoring methods, right? I think this was a simplification because it doesn't use probability=True which makes the SVM use platt scaling which is a pretty weird thing to do, imho, in particular inside the cross-validation.

The problem was that with the default tolerance there are no digits that are equal, I think.

twosigmajab committed:
DOC Add libraries.io and changelog links (#11298)

jeremiedbb committed:
DOC reorganize datasets documentation page (#11180)

georgipeev commented:
"All checks have passed" - so elusive, but apparently still attainable.

TomDLT commented on sklearn/svm/classes.py:
```diff
@@ -352,7 +351,7 @@ class LinearSVR(LinearModel, RegressorMixin):
     sklearn.linear_model.SGDRegressor
         SGDRegressor can optimize the same cost function as LinearSVR
-        by adjusting the penalty and loss parameters. In addition it requires
+        by adjusting the penalty and loss parameters. In addition it .requires
```
mystake

Commit:
removed extra dot

qinhanmin2014 commented:
> The problem was that with the default tolerance there are no digits that are equal, I think.

@georgipeev Could you take some time to confirm? I would be surprised if this is the case. Thanks in advance.

georgipeev commented:
>The problem was that with the default tolerance there are no digits that are equal, I think.

@amueller's recollection is correct - at the default tolerance many values in these examples only matched 1-2 digits after the decimal point, and one value matched none when running against whichever RNG library gets used under OS X.

qinhanmin2014 commented:
> @amueller's recollection is correct - at the default tolerance many values in these examples only matched 1-2 digits after the decimal point, and one value matched none when running against whichever RNG library gets used under OS X.

Thanks. I think I'll vote +0 here, since I still doubt whether it's good to make examples complex. E.g., I think ``clf = LinearSVC(random_state=0)`` is a good example for new comers, but ``clf = LinearSVC(random_state=0, tol=0.00001)`` is not.
Maybe we have some other ways, e.g., modifying ``n_features`` or ``random_state`` in ``make_classification(n_features=4, random_state=0)``.
Anyway, thanks @georgipeev for the PR. I'm fine with merging if we have two approvals here.

georgipeev commented:
CI failure seems unrelated. How should we proceed?

qinhanmin2014 commented:
@georgipeev Master branch is failing now. Please merge master branch in after #11318 is merged.

mfeurer committed:
ENH Add refit_time_ attribute to model selection (#11310)

markroth8 committed:
MAINT Fix #9350: Enable has_fit_parameter() and fit_score_takes_y() to work with @deprecated in Python 2 (#11277)

jnothman committed:
Fix skipping in conftest.py (#11318)

jnothman committed:
MAINT clarifications in ColumnTransformer._update_transformers (#11323)

jnothman committed:
COSMIT fix syntax quirk

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge branch 'master' into fix-doctest-for-LinearSVC-and-LinearSVR

georgipeev commented:
Is anything else required before this PR can be merged?

amueller commented:
Something weird happened with your history but I think it'll go away if we squash and merge. looks good otherwise.

glemaitre committed:
nitpicks

glemaitre commented:
LGTM merging.
Thanks @georgipeev

TheAtomicOption commented:
I tested this issue **before the above merge** with a conda env that included mkl in an attempt to get the same result on macOS, but the test still failed. While the above merge fixes the tests by removing precision, it doesn't align the actual results. 

The liblinear code is directly compiled against the system blas as in the output below. This likely explains the system specific discrepancy. 
```
sklearn/svm/src/liblinear/tron.cpp:16:10:
#include <cblas.h>
```

amueller commented:
@TheAtomicOption yes, so there's no way to align the results without increasing the precision a lot. You can increase the precision and align the results but we found that wasn't worth it for a doctest.
